Using the Debugger (#67)
Ada-Developers-Academy/core-unit-1: intermediate-python-debugging/vs-code-debugging.md

> # Debugging With VS Code

Reviewer: My ideal is either:

"Debugging with VS Code" and file name `debugging-with-vs-code`
or "VS Code Debugging" and file name `vs-code-debugging`

Author: Done

Python is widely used with VS Code and so VS Code's developers have put the time in to integrate the Python debugger into the standard VS Code plugin for Python.

In this lesson we will learn to connect VS Code and the Python debugger to help us fix errors in our code.

> ## Learning Goals
> 
> By the end of this lesson we should be able to...
> 
> - Set up VS Code to run the Python debugger
> - Run the Python debugger in VS Code to step through a program
> - Use breakpoints, and watches to identify errors in code

Reviewer: Another suggestion I have is to convert this into a "Goals and Structure" section. This will establish what the lesson is trying to do and how. This is useful to me because this lesson reads way more like a directed/guided exercise with steps in order (with a lot of "you do this now, observe this now, answer this now"), rather than disseminating knowledge.

```suggestion
## Goal and Structure

In order to use a debugger in VS Code while working on Python projects, we need to set up our tools and then learn how to navigate and use them.

This lesson is broken up into the following sections:

1. Setting up VS Code to use and run the Python debugger
2. Using a small example Python program to learn how to and practice:
    - Adding and removing breakpoints
    - Running the debugger and watching the program pause
    - Observing variable state at a breakpoint
    - Creating and using a watch
    - Resuming, stepping over, stepping into, step out, restart, and stop
```

## Before Starting

Create a new project folder and Python file.

```
$ mkdir testing-example # make a testing-example folder
$ cd testing-example    # change into that folder
$ touch guess_number.py # create an empty python fole
$ code .                # Open the folder as a project in VS Code
```

Fill `guess_number.py` with the following python code.

```python
from random import randrange

number = randrange(10)

guess = -1
while number != guess:
  guess = int(input('Please enter a guess between 0-10  ==>'))
  if guess < number:
    print(f"{guess} is too low")
  elif number > guess:
    print(f"{guess} is too high")
  else:
    print("You guessed it!")

```

## Setting Up

To get started with the VS Code debugger, first find the debugger button on the left sidebar.

![Debugger Icon](../assets/vs-code-debugger/debugger-icon.png)

You may then need to create a `launch.json` file by clicking on `Run and Debug`.

![Run and debug button](../assets/vs-code-debugger/run-and-debug.png)

> Then Select `create a launch.json file` To make a configuration.  

Reviewer: ```suggestion
Then select `create a launch.json file` to make a configuration.  
```

![Debug Python File](../assets/vs-code-debugger/debug-configuration.png)

> Save the resulting `launch.json` file.
> 
> That's it!  You are set up!  You can close `launch.json` now.

Reviewer: Move away from "That's it!" tone

```suggestion
Save the resulting `launch.json` file.

You are set up!  You can close `launch.json` now.
```

## Debugging

> ### Adding Breakpoints

Reviewer: Because you mention toggling:

```suggestion
### Adding and Removing Breakpoints
```

You can add a breakpoint by clicking to the **left** of the line number.  You will see a red dot appear to indicate the breakpoint.  You can click again to make it dissapear.  

![Breakpoint](../assets/vs-code-debugger/breakpoint.png)

Add a breakpoint to lines 5 and 11.  See how you can toggle them on and off.

### Running the Debugger

You can then run the debugger at any time with the green triangle play button.

![Debugger play button](../assets/vs-code-debugger/debugger-play.png)

> Notice that the program will pause at the 1st breakpoint it encounters.

Reviewer: ```suggestion
Notice that the program will pause at the first breakpoint it encounters.
```

**What are the values of the variables?**  

### Watches

You can create a watch by clicking on the `+` icon and enter any expression.

![Entering a Watch](../assets/vs-code-debugger/watch-price.png)

![Watches](../assets/vs-code-debugger/watch.png)

> **In the watch box, enter `number < guess`**  What does it evaluate to?  Try a few more expressions.

Reviewer: ```suggestion
**In the watch box, enter `number < guess`**.  What does it evaluate to?  Try a few more expressions.
```

![Watch example](../assets/vs-code-debugger/watch_example.png)

### Stepping Over & Stepping Into

Notice when the debugger is paused at a breakpoint you get the menu below.

![Debugger Menu](../assets/vs-code-debugger/debugger-menu.png)

This menu has the following buttons:

- **Resume** - Continues the debugger executing until it gets to the next breakpoint, or the program exits.
- **Step Over** - This button executes the current line and moves to the next line.  It will execute any function invoked on the current line, without pausing inside that function.
- **Step Into** - This button goes into any function invoked on the current line and pauses at the 1st line.
- **Step Out** - This button finishes the rest of the current function and pauses at the calling function.
- **Restart** - This button restarts the debugger from the start.

> - **Stop** - This button stops the debugger.

Reviewer: Is it worth it to turn this into a table instead?

<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
<!-- Replace everything in square brackets [] and remove brackets  -->

### !challenge

* type: short-answer
* id: c3eb6210-5cbd-4507-b0a5-41db00b43fb5
* title: Step over and Step into why no difference?
* points: [1] (optional, the number of points for scoring as a checkpoint)
<!-- * topics: [python, pandas] (optional the topics for analyzing points) -->

##### !question

In the code try using the step over and step into buttons, do you notice a difference?  No, why not?
##### !end-question

##### !placeholder

Why don't we see a difference?

##### !end-placeholder

##### !answer

/.+/

##### !end-answer

<!-- other optional sections -->
<!-- !hint - !end-hint (markdown, hidden, students click to view) -->
<!-- !rubric - !end-rubric (markdown, instructors can see while scoring a checkpoint) -->
##### !explanation 

It's because we don't have any functions that we have written and used  The debugger is wisely deciding not to jump into the source code for the `input` function.

##### !end-explanation

### !end-challenge

> <!-- ======================= END CHALLENGE ======================= -->

Reviewer: Personally, I don't think this is an effective question for students to spend their time on.

The student takeaway should be that there **is** a difference between step over and step in buttons. With this question, the student must put effort into understanding this exception. Since this is their first time with the subject, the exception may stick out more clearly in their head than the rule, which isn't ideal IMO.

Personally, I'd either cut this question, or turn this question into assessing their knowledge of something more "typical" (and possibly move into a Check for Understanding section)

> ## Summary
> 
> In this lesson we examined how to get set up using the debugger in VS Code.  We also learned to create breakpoints watch programmatic expressions and trace through the application using step over, step into and step out.  

Reviewer: In my opinion, this Summary section doesn't add much to the lesson. The curriculum thus far has used Summary sections to summarize large lessons with multiple things going on, and it's left out summary sections on smaller "more straightforward" lessons. I'd suggest the same here

```suggestion
```

> 

Reviewer: Conform to other lessons by having a Check for Understanding header

```suggestion
## Check for Understanding
```
